### ISSUE TITLE
Issue with rosbag2_transport Recorder - not continuously recording after node start

### ISSUE THREAD

**Issue by cnicho35** [ros2/rosbag2#1696]
## Description
Not necessarily a bug, but I could use some help making the rosbag2_transport recorder work.

### Expected Behavior
The recorder should be started by this c++ code (implemented in a ros2 node). It should record by specified topics as mcap, with a file duration limit (specified in a parameter).

### Actual Behavior
The node compiles, executes, successfully subscribes to my specified topic but it does not capture any messages. The mcap file remains at 0 bytes until I ctrl+c the ros2 launch file. It then properly closes the bag file and a metadata.yaml file is created but it shows that 0 messages were recorded.

## To Reproduce
```
#include <rclcpp/rclcpp.hpp>
#include <rosbag2_cpp/writer.hpp>
#include <rosbag2_transport/recorder.hpp>
#include <rosbag2_transport/record_options.hpp>
#include <rosbag2_storage/storage_options.hpp>
#include <chrono>
#include <memory>
#include <vector>
#include <string>
#include <iostream>
#include <iomanip>
#include <ctime>

class BagRecorder : public rclcpp::Node
{
public:
  BagRecorder()
      : Node("bag_recorder")
  {
    // Declare parameters with default values
    this->declare_parameter<std::string>("dataFolder", "/home/user/Data");
    this->declare_parameter<int>("fileDuration", 60);
    this->declare_parameter<std::vector<std::string>>("loggedTopics", {"/rosout", "/system_messenger", "/labjack_ain"});

    // Get parameter values
    data_folder_ = this->get_parameter("dataFolder").as_string();
    file_duration_ = this->get_parameter("fileDuration").as_int();
    logged_topics_ = this->get_parameter("loggedTopics").as_string_array();

    // Log parameter values for verification
    std::cout << "Data Folder: " << data_folder_ << std::endl;
    std::cout << "File Duration: " << file_duration_ << " seconds" << std::endl;
    std::cout << "Logged Topics: ";
    for (const auto& topic : logged_topics_)
    {
      std::cout << topic << " ";
    }
    std::cout << std::endl;

    // Set storage options
    storage_options_ = {
        getDataPath(data_folder_), "mcap", 0, static_cast<uint64_t>(file_duration_), 100
    };

    std::cout << "Storage Path: " << storage_options_.uri << std::endl;

    writer_ = std::make_shared<rosbag2_cpp::Writer>();

    // Set record options
    rosbag2_transport::RecordOptions record_options = {
        false, false, logged_topics_, "cdr", std::chrono::milliseconds(1000)
    };

    recorder_ = std::make_shared<rosbag2_transport::Recorder>(
        std::move(writer_), storage_options_, record_options);

    // Start recording
    start_recording();
  }

  void start_recording()
  {
    recorder_->record();
  }

  void stop_recording()
  {
    recorder_->stop();
  }

  ~BagRecorder()
  {
    stop_recording();
  }

private:
  std::string data_folder_;
  int file_duration_;
  std::vector<std::string> logged_topics_;
  rosbag2_storage::StorageOptions storage_options_;
  std::shared_ptr<rosbag2_transport::Recorder> recorder_;
  std::shared_ptr<rosbag2_cpp::Writer> writer_;

  std::string getCurrentUtcTime()
  {
    auto now = std::chrono::system_clock::now();
    std::time_t now_time_t = std::chrono::system_clock::to_time_t(now);
    std::tm utc_tm = *std::gmtime(&now_time_t);
    std::ostringstream oss;
    oss << std::put_time(&utc_tm, "%Y_%m_%d_%H_%M_%S");
    return oss.str();
  }

  std::string getDataPath(const std::string &baseFolder)
  {
    std::string systemStartTime = getCurrentUtcTime();
    return baseFolder + "/Bag_" + systemStartTime;
  }
};

int main(int argc, char *argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<BagRecorder>();

  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

```

Terminal Output:
```
[INFO] [launch]: All log files can be found below /home/user/.ros/log/2024-06-06-15-09-46...
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [labjack_ain_reader-3]: process started with pid [487262]
[INFO] [bag_recorder_node-5]: process started with pid [487277]
[bag_recorder_node-5] Data Folder: /home/user/Data
[bag_recorder_node-5] File Duration: 60 seconds
[bag_recorder_node-5] Logged Topics: /rosout /system_messenger /labjack_ain 
[bag_recorder_node-5] Storage Path: /home/user/Data/Bag_2024_06_06_21_09_46
[bag_recorder_node-5] stdin is not a terminal device. Keyboard handling disabled.[WARN] [1717708186.930026264] [rcl.logging_rosout]: Publisher already registered for provided node name. If this is due to multiple nodes with the same name then all logs for that logger name will go out over the existing publisher. As soon as any node with that name is destructed it will unregister the publisher, preventing any further logs for that name from being published on the rosout topic.
[bag_recorder_node-5] [INFO] [1717708186.946501207] [BagRecorder]: Press SPACE for pausing/resuming
[bag_recorder_node-5] [INFO] [1717708186.949672648] [BagRecorder]: Listening for topics...
[bag_recorder_node-5] [INFO] [1717708186.952973855] [BagRecorder]: Subscribed to topic '/system_messenger'
[bag_recorder_node-5] [INFO] [1717708186.953967506] [BagRecorder]: Event publisher thread: Starting
[bag_recorder_node-5] [INFO] [1717708186.954659994] [BagRecorder]: Subscribed to topic '/rosout'
[bag_recorder_node-5] [INFO] [1717708186.954815241] [BagRecorder]: Recording...
[bag_recorder_node-5] [INFO] [1717708187.957239015] [BagRecorder]: Subscribed to topic '/labjack_ain'
[bag_recorder_node-5] [INFO] [1717708187.957311422] [BagRecorder]: All requested topics are subscribed. Stopping discovery...
^C[WARNING] [launch]: user interrupted with ctrl-c (SIGINT)
[bag_recorder_node-5] [INFO] [1717708210.366576834] [rclcpp]: signal_handler(signum=2)
[labjack_ain_reader-3] [INFO] [1717708210.366860781] [rclcpp]: signal_handler(signum=2)
[bag_recorder_node-5] [INFO] [1717708210.380178365] [rosbag2_cpp]: Writing remaining messages from cache to the bag. It may take a while
[bag_recorder_node-5] [INFO] [1717708210.384200159] [BagRecorder]: Event publisher thread: Exiting
[bag_recorder_node-5] [INFO] [1717708210.384547453] [BagRecorder]: Recording stopped```
```

## System (please complete the following information)
 - OS: Ubuntu 22.04
 - ROS 2 Distro: Humble
 - Install Method: APT
 - Version: release

## Additional context
Any help would be greatly appreciated!!!

The reason I am not using the standard _ros2 bag record_ functionality is because I want programmatic control over the start and stop of bag file recording (more than the pause ability).

**Comment by cnicho35** [comment]
Made some progress I wanted to share:

Needed to add a multithread executor to start the rosbag2_recorder node from the BagRecorder management node.

Unfortunately, I still have a problem that I cannot stop the recorder and restart it with a new folder path. Any ideas how to implement this?

I tried adding a function that stops the records and reinitializes it but this doesn't stop the old node running on a separate thread. When I run it, it does the same problem where it creates a new blank bag file but doesn't write to it. I assume the issue is caused by a node with a duplicate name.

code:
```
void reset_bag_recording(const std::string &new_data_folder)
  {
    stop_recording();
    
    data_folder_ = new_data_folder;
    storage_options_.uri = getDataPath(data_folder_);
    writer_ = std::make_shared<rosbag2_cpp::Writer>();

    rosbag2_transport::RecordOptions record_options = {
        false, false, logged_topics_, "cdr", std::chrono::milliseconds(1000)};

    recorder_ = std::make_shared<rosbag2_transport::Recorder>(
        writer_, storage_options_, record_options);

    std::cout << "New Storage Path: " << storage_options_.uri << std::endl;

    start_recording();
  }
```

...

```
int main(int argc, char *argv[])
{
  rclcpp::init(argc, argv);
  auto bag_recorder_node = std::make_shared<BagRecorder>();

  // Use a MultiThreadedExecutor to spin the BagRecorder and Recorder nodes
  rclcpp::executors::MultiThreadedExecutor executor;
  executor.add_node(bag_recorder_node);
  executor.add_node(bag_recorder_node->get_recorder());

  // Example of how to reset the recorder with a new folder path
  std::thread([bag_recorder_node]() {
    std::this_thread::sleep_for(std::chrono::seconds(10));
    bag_recorder_node->reset_bag_recording("/new/data/folder/path");
  }).detach();

  executor.spin();

  rclcpp::shutdown();
  return 0;
```

Working code without stop and restart ability:

```
#include <rclcpp/rclcpp.hpp>
#include <rosbag2_cpp/writer.hpp>
#include <rosbag2_transport/recorder.hpp>
#include <rosbag2_transport/record_options.hpp>
#include <rosbag2_storage/storage_options.hpp>
#include <chrono>
#include <memory>
#include <vector>
#include <string>
#include <iostream>
#include <iomanip>
#include <ctime>
using namespace std::chrono_literals;

class BagRecorder : public rclcpp::Node
{
public:
  BagRecorder()
      : Node("BagRecorder")
  {
    // Declare parameters with default values
    this->declare_parameter<std::string>("dataFolder", "/home/user/Data");
    this->declare_parameter<int>("fileDuration", 60);
    this->declare_parameter<std::vector<std::string>>("loggedTopics", {"/rosout", "/system_messenger", "/labjack_ain"});

    // Get parameter values
    data_folder_ = this->get_parameter("dataFolder").as_string();
    file_duration_ = this->get_parameter("fileDuration").as_int();
    logged_topics_ = this->get_parameter("loggedTopics").as_string_array();
    // Log parameter values for verification
    std::cout << "Data Folder: " << data_folder_ << std::endl;
    std::cout << "File Duration: " << file_duration_ << " seconds" << std::endl;
    std::cout << "Logged Topics: ";
    for (const auto &topic : logged_topics_)
    {
      std::cout << topic << " ";
    }
    std::cout << std::endl;

    // Set storage options
    storage_options_.uri = getDataPath(data_folder_);
    storage_options_.storage_id = "mcap";
    storage_options_.max_bagfile_size = 0;
    storage_options_.max_bagfile_duration = std::chrono::seconds(file_duration_).count();
    storage_options_.max_cache_size = 100;
    storage_options_.storage_preset_profile = "";
    storage_options_.snapshot_mode = false;

    std::cout << "Storage Path: " << storage_options_.uri << std::endl;

    writer_ = std::make_shared<rosbag2_cpp::Writer>();

    // Set record options
    rosbag2_transport::RecordOptions record_options = {
        false, false, logged_topics_, "cdr", std::chrono::milliseconds(1000)};

    // Initialize recorder with unique node name
    recorder_ = std::make_shared<rosbag2_transport::Recorder>(
        writer_, storage_options_, record_options);

    // Start recording
    start_recording();

  }

  void start_recording()
  {

    recorder_->record();
  }

  void stop_recording()
  {
    recorder_->stop();
  }



  ~BagRecorder()
  {
    stop_recording();
  }

  std::shared_ptr<rosbag2_transport::Recorder> get_recorder() const
  {
    return recorder_;
  }

private:
  std::string data_folder_;
  int file_duration_;
  std::vector<std::string> logged_topics_;
  rosbag2_storage::StorageOptions storage_options_;
  rclcpp::NodeOptions node_options_;
  std::shared_ptr<rosbag2_transport::Recorder> recorder_;
  std::shared_ptr<rosbag2_cpp::Writer> writer_;

  std::string getCurrentUtcTime()
  {
    auto now = std::chrono::system_clock::now();
    std::time_t now_time_t = std::chrono::system_clock::to_time_t(now);
    std::tm utc_tm = *std::gmtime(&now_time_t);
    std::ostringstream oss;
    oss << std::put_time(&utc_tm, "%Y_%m_%d_%H_%M_%S");
    return oss.str();
  }

  std::string getDataPath(const std::string &baseFolder)
  {
    std::string systemStartTime = getCurrentUtcTime();
    return baseFolder + "/Bag_" + systemStartTime;
  }
};

int main(int argc, char *argv[])
{
  rclcpp::init(argc, argv);
  auto bag_recorder_node = std::make_shared<BagRecorder>();

  // Use a MultiThreadedExecutor to spin the BagRecorder and Recorder nodes
  rclcpp::executors::MultiThreadedExecutor executor;
  executor.add_node(bag_recorder_node);
  executor.add_node(bag_recorder_node->get_recorder());

  executor.spin();

  rclcpp::shutdown();
  return 0;
}
```

**Comment by cnicho35** [comment]
Working as expected now. Still welcoming better methods if this seems incorrect to you. I'm no c++ expert...
Closes the old bag file and reopens one every 10 minutes.

Closing with working code:

```
#include <rclcpp/rclcpp.hpp>
#include <rosbag2_cpp/writer.hpp>
#include <rosbag2_transport/recorder.hpp>
#include <rosbag2_transport/record_options.hpp>
#include <rosbag2_storage/storage_options.hpp>
#include <chrono>
#include <memory>
#include <vector>
#include <string>
#include <iostream>
#include <iomanip>
#include <ctime>
using namespace std::chrono_literals;

class BagRecorder : public rclcpp::Node
{
public:
  BagRecorder()
      : Node("BagRecorder")
  {
    // Declare parameters with default values
    this->declare_parameter<std::string>("dataFolder", "/home/user/Data");
    this->declare_parameter<int>("fileDuration", 60);
    this->declare_parameter<std::vector<std::string>>("loggedTopics", {"/rosout", "/system_messenger", "/labjack_ain"});

    // Get parameter values
    data_folder_ = this->get_parameter("dataFolder").as_string();
    file_duration_ = this->get_parameter("fileDuration").as_int();
    logged_topics_ = this->get_parameter("loggedTopics").as_string_array();

    reset_bag_timer = this->create_wall_timer(600s, std::bind(&BagRecorder::reset_bag_recording, this));
  }

  void start_recording()
  {

    // Log parameter values for verification
    std::cout << "Data Folder: " << data_folder_ << std::endl;
    std::cout << "File Duration: " << file_duration_ << " seconds" << std::endl;
    std::cout << "Logged Topics: ";
    for (const auto &topic : logged_topics_)
    {
      std::cout << topic << " ";
    }
    std::cout << std::endl;

    // Set storage options
    storage_options_.uri = getDataPath(data_folder_);
    storage_options_.storage_id = "mcap";
    storage_options_.max_bagfile_size = 0;
    storage_options_.max_bagfile_duration = std::chrono::seconds(file_duration_).count();
    storage_options_.max_cache_size = 100;
    storage_options_.storage_preset_profile = "";
    storage_options_.snapshot_mode = false;

    std::cout << "Storage Path: " << storage_options_.uri << std::endl;

    writer_ = std::make_shared<rosbag2_cpp::Writer>();

    // Set record options
    rosbag2_transport::RecordOptions record_options = {
        false, false, logged_topics_, "cdr", std::chrono::milliseconds(1000)};

    // Initialize recorder with unique node name
    recorder_ = std::make_shared<rosbag2_transport::Recorder>(
        writer_, storage_options_, record_options);

    // Start recording
    executor->add_node(recorder_);
    recorder_->record();
  }

  void stop_recording()
  {
    recorder_->stop();
    executor->remove_node(recorder_);
  }

  void reset_bag_recording()
  {
    stop_recording();

    start_recording();
  }

  ~BagRecorder()
  {
    stop_recording();
  }

  std::shared_ptr<rosbag2_transport::Recorder> get_recorder() const
  {
    return recorder_;
  }

  void set_executor(std::shared_ptr<rclcpp::executors::MultiThreadedExecutor> executorSet)
  {
    executor = executorSet;
  }

private:
  rclcpp::TimerBase::SharedPtr reset_bag_timer;

  std::string data_folder_;
  std::shared_ptr<rclcpp::executors::MultiThreadedExecutor> executor;
  int file_duration_;
  std::vector<std::string> logged_topics_;
  rosbag2_storage::StorageOptions storage_options_;
  rclcpp::NodeOptions node_options_;
  std::shared_ptr<rosbag2_transport::Recorder> recorder_;
  std::shared_ptr<rosbag2_cpp::Writer> writer_;

  std::string getCurrentUtcTime()
  {
    auto now = std::chrono::system_clock::now();
    std::time_t now_time_t = std::chrono::system_clock::to_time_t(now);
    std::tm utc_tm = *std::gmtime(&now_time_t);
    std::ostringstream oss;
    oss << std::put_time(&utc_tm, "%Y_%m_%d_%H_%M_%S");
    return oss.str();
  }

  std::string getDataPath(const std::string &baseFolder)
  {
    std::string systemStartTime = getCurrentUtcTime();
    return baseFolder + "/Bag_" + systemStartTime;
  }
};

int main(int argc, char *argv[])
{
  rclcpp::init(argc, argv);
  auto bag_recorder_node = std::make_shared<BagRecorder>();

  // Use a MultiThreadedExecutor to spin the BagRecorder and Recorder nodes
  auto executor = std::make_shared<rclcpp::executors::MultiThreadedExecutor>();

  executor->add_node(bag_recorder_node);
  bag_recorder_node->set_executor(executor);
  bag_recorder_node->start_recording();
  // Example of how to reset the recorder with a new folder path
  //  std::thread([bag_recorder_node]() {
  //   std::this_thread::sleep_for(std::chrono::seconds(10));
  //   bag_recorder_node->reset_bag_recording();
  // }).detach();

  executor->spin();

  rclcpp::shutdown();
  return 0;
}
```
